### PR TITLE
Modify the windows native build to be razzle friendly

### LINKED ISSUE #5854: Native ClrCompression.dll for win7-x64 package depends on MSVCR110.dll
@ianhays :

Much like issue #3760 for SqlClient, the 64-bit specific version of clrcompression.dll takes a dependency on MSVCR110.dll.  The issue appears to be specific to the x64 version currently.

### PR DESCRIPTION
**Pull request by ianhays**
Internally we are using an existing vcproj to build the native windows binaries. This commit modifies the CMake build script in the open to allow more fine grained control of the build environment. The result of this is that we can specify the necessary environment args from the caller, enabling us to run the build-native script from fxcore/src in a Razzle environment via a wrapper project.

resolves #5854 when accompanying TFS commit goes through

@weshaggard @ericstj

### CODE REVIEW

**Comment by ericstj** [comment]
Looks ok.  Ideally I'd like it if the open build also defaulted to build against the OS CRT so that when we eventually ditch TFS we don't have to remember this issue so as not to regress.  Is there a way to make that happen?  Curious what https://github.com/dotnet/coreclr does for this problem.  /cc @jkotas

**Comment by ianhays** [comment]
>defaulted to build against the OS CRT

With the -LinkLibrary and LinkArgument tags we could modify the calling source in build.cmd to reference the OS CRT. Or we could make a clrcompression wrapper in open with the desired args. I'd prefer to leave build-native.cmd as modular as possible for other potential native libraries, cross-compilation capability, and/or building without requiring VS, so I'd prefer to keep the library specification outside of the script.

The default CMake VS generator behavior links the resulting clrcompression.dll against VCRuntime140.dll and Kernel32.dll. It uses vcvarsall.bat.

**Comment by jkotas** [comment]
> Curious what https://github.com/dotnet/coreclr does for this problem. /cc @jkotas 

The internal builds are linking against OS CRT. The external builds are linking against regular VS CRT.

@gkhanna79 has been working on a plan to move both to CoreSystem-compatible UCRT.

**Comment by ericstj** [comment]
Thanks @jkotas.  In that case this appears to be consistent with coreclr.  Let's make sure @gkhanna79  knows that we'll need to update corefx to UCRT once that work happens.

:shipit:

**Comment by gkhanna79** [comment]
For CoreCLR, I am working on the following:

1) https://github.com/dotnet/coreclr/issues/3014
2) https://github.com/dotnet/coreclr/issues/3013

Both these need to happen for CoreFX as well.

File: src/Native/Windows/build-native.cmd [hunk 1]
```diff
@@ -10,36 +10,53 @@ set __IntermediatesDir=""
 set __BuildArch=AnyCPU
```
Comment (weshaggard) [inline src/Native/Windows/build-native.cmd]: BuildArch AnyCPU doesn't seem right in any scenario for native built binaries.

File: src/Native/Windows/build-native.cmd [hunk 2]
```diff
@@ -10,36 +10,53 @@ set __IntermediatesDir=""
...
+    if /i [%2] == [amd64]           ( set __BuildArch=x64&&set __VCBuildArch=x86_amd64&&shift&&shift&goto Arg_Loop)
     echo Error: Invalid platform args "%1 and %2"
     exit /b 1
 )
-if /i [%1] == [-intermediateDir]    (
-    set __IntermediatesDir=%2
-)
-if /i [%1] == [-binDir]    (
-    set __CMakeBinDir=%2
+if /i [%1] == [-intermediateDir]    ( set __IntermediatesDir=%2&&shift&&shift&goto Arg_Loop)
```
Comment (weshaggard) [inline src/Native/Windows/build-native.cmd]: For these arguments I assume you cannot pass them through build.cmd because they would cause the managed build to fail. Do you know if any of these are actually used? If not we should remove them. If they are we should consider having them follow the same pattern as the other options.

File: src/Native/Windows/build-native.cmd [hunk 3]
```diff
@@ -10,36 +10,53 @@ set __IntermediatesDir=""
 set __BuildArch=AnyCPU
 set __VCBuildArch=x86_amd64
 set CMAKE_BUILD_TYPE=Release
+set __EnvironmentSet=false
+set __VSVersion=vs2013
```
Comment (weshaggard) [inline src/Native/Windows/build-native.cmd]: Why are we picking VS2013 version? For the managed side we have removed support for VS2013 so we should do the similar for the native side.

Comment (weshaggard) [inline src/Native/Windows/build-native.cmd]: Also at one point we talked about having cmake stop generating VS solutions for windows and have it consistent for both windows and x-plat and just using make. Is that something you have looked into yet?

File: src/Native/Windows/build-native.cmd (hunk 1 above)
Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: It isn't. It's really a default to amd64. See line 114.

File: src/Native/Windows/build-native.cmd (hunk 3 above)
Comment (jkotas) [inline src/Native/Windows/build-native.cmd]: What would we get by using make instead of msbuild?

It would add one more pre-req for building on Windows...

Comment (weshaggard) [inline src/Native/Windows/build-native.cmd]: We already have a dependency on cmake because of this. 

The biggest thing we would get is a consistent build between the platforms. I don't know if there are any issues with between the differences between make and msbuild generates stuff but in general I'm trying to cut down on the differences between our different build environments. Do we get any benefits to having it generate msbuild files? The only one I can see is being able to load these projects in VS but most of these are very small native libraries so that benefit seems minimal to me.

Comment (jkotas) [inline src/Native/Windows/build-native.cmd]: I believe that CMake saves you from maintaining boring stuff: detection of where things are, reasonable defaults, ... . make would give the same build driver, but you pay for it by having more platform-specific plumbing that has be maintained.

Comment (weshaggard) [inline src/Native/Windows/build-native.cmd]: Sorry reading my earlier comment I think there is some miss-understanding. I wasn't implying we stop using CMake I was just implying that we should use CMake the same way on windows and linux/osx. Honestly I'm not entirely sure how this works on linux/osx (I should dig into this more) but all I see is that we have a bunch of extra scripts specific to windows which I would like to eliminate if possible.

File: src/Native/Windows/build-native.cmd (hunk 2 above)
Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: >For these arguments I assume you cannot pass them through build.cmd because they would cause the managed build to fail.

Yes. The arguments that use a hyphen were done so because I didn't think they would be generally passed to build-native through build.cmd.

>Do you know if any of these are actually used?

The args with the hyphens are all used by the internal wrapper proj that is in my Codeflow review.

>If they are we should consider having them follow the same pattern as the other options.

We could certainly do that. I went with this format because I didn't want to potentially overwrite a property that has an existing meaning already.If I prepend "native" to the names then that worry should be gone, so I'll do that.

File: src/Native/Windows/build-native.cmd (hunk 3 above)
Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: Some general background here to frame the discussion: CMake requires the use of a "generator" to produce the intermediary files that can then be compiled to a dll or lib file. On Unix, the generator is Clang and it produces a MakeFile. On Windows, the generator is Visual Studio and it produces a vcxproj.

The Unix and Windows native builds therefore have two steps:
- Call CMake on the sources to produce the intermediate files using clang (unix) or VS (windows) as the generator
- Call Make (unix) or msbuild (Windows) on the intermediate files to produce the native binary

>we should use CMake the same way on windows and linux/osx

The goal I'm moving towards is to enable cross-compilation of CMake such that the native binaries for Unix may be built on Windows and vice-versa, but building them the <b>same</b> way is a bit more complicated. The main issue blocking this is the lack of a stable generator that is compatible with both platforms; clang for Windows is still early stages, and msbuild on Unix may not adequately replace full VS enough to qualify as a generator. I really need to delve deeper into Clang for Windows since that's likely our best option for building the same way, but I haven't a chance to do so yet.

If we just wanted to produce the same <i>type</i> of intermediate file then we could use nmake as the generator instead of VS for Windows, but that still isn't the same method as Unix since the generator would be different.

>but all I see is that we have a bunch of extra scripts specific to windows which I would like to eliminate if possible.

That's my eventual goal after cross-compilation support. I'd like to have the logic for the Unix build the same as the Windows logic, generator and all, but it will take some extra work that is mostly orthogonal to getting the internal build and open build at parity.

>Also at one point we talked about having cmake stop generating VS solutions for windows and have it consistent for both windows and x-plat and just using make. Is that something you have looked into yet?

In short: yes I have looked into it, but not enough to be ready to move forward with.

Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: > Why are we picking VS2013 version? For the managed side we have removed support for VS2013 so we should do the similar for the native side.

This is the default, but it may as well be set to an empty string. It is in the :ToolsVersion tag that we set the actual value via a check of the environment variables VisualStudioVersion/VS140COMNTOOLS/VS120COMNTOOLS. 

We use the higher version available but I can remove VS2013 support if that's the way of the managed side.

Comment (jkotas) [inline src/Native/Windows/build-native.cmd]: Ideally, we should be keeping these things similar where possible between the different dotnet repos. Both coreclr and corert have `set __VSVersion=vs2015`.

Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: Thanks @jkotas, I'll change it to vs2015.

At some point we should look into consolidating the cmake bits between coreclr and corefx a la https://github.com/dotnet/coreclr/issues/1344. The CoreFX and CoreCLR windows native builds used to use the exact same code, so we could certainly get back to that point again in the future. That's a discussion we can continue in that issue, though.

Comment (jkotas) [inline src/Native/Windows/build-native.cmd]: > The CoreFX and CoreCLR windows native builds used to use the exact same code

As far as I know, it was never the case. There was always repo-specific code for the native builds, but the stuff that could have been same was actually same.

Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: >There was always repo-specific code for the native builds, but the stuff that could have been same was actually same.

You are correct; I should not have said exact same. Although I directly copied the CoreCLR native build logic into CoreFX, I did have to make a few modifications that were CoreFX specific.

What I should have said was that there was enough shared code between CoreCLR and CoreFX at the inception of the CoreFX native windows build that we could have relatively easily combined the logic into one set without adding too much complexity.

**Comment by ianhays** [comment]
Closing this and doing the upgrade on the TFS side.
